docs/it/design-pattern.md

```diff
@@ -0,0 +1,126 @@
+
+In generale, design patterns e programmazione ad oggetti sono strettamente legati poiché spesso i primi vengono descritti e spiegati utilizzando di oggetti.
+Questo non vuol dire che, ad esempio, nella programmazione funzionale non vengano utilizzati, ma naturalmente necessitano di strutture differenti; in alcuni casi, a dire il vero, alcuni Design Patterns sono più facilmente implementabili con quest'ultimo paradigma.
+
+Internet ha molte risorse valide sull'argomento, fra queste possiamo citare:
+
+- [Refactoring Guru](https://refactoring.guru/design-patterns/) molto dettagliato e che spiega ogni Design Pattern evidenziando di ognuno vantaggi e svantaggi e confrontandoli con gli altri. Sono presenti esempi in codice acquistabili separatamente in tutti i linguaggi.
+- [Il canale Youtube di Christopher Okhravi](https://www.youtube.com/@ChristopherOkhravi)
+- [Design Patterns su Wikipedia Italiano](https://it.wikipedia.org/wiki/Design_pattern)
+- [Design Patterns su Wikipedia Inglese](https://en.wikipedia.org/wiki/Design_pattern)
+
+Il libro della Gang of Four identifica un elenco iniziale di design pattern, ognuno dei quali ha un nome che identifica la soluzione al problema generico che andrà a risolvere.
+Generalmente chi sta sviluppando una certa porzione di codice per implementare una o più funzionalità potrebbe "riconoscere" un design pattern durante l'analisi di un problema specifico che si sta cercando di risolvere e decidere di applicarlo, ottenendo una serie di vantaggi, tra cui:
+
+- **Facilità di riuso e test del software**: Ogni design pattern è stato studiato e catalogato di modo da essere una soluzione pressoché universale ad ogni problema analogo a quello descritto, di conseguenza è possibile riutilizzare una implementazione in più punti del proprio codice. Inoltre, questi pattern sono solitamente accompagnati da suggerimenti e metodologie per il loro testing, il che rende il processo di scrittura dei test automatizzati più semplice ed efficace.
+- **Miglior comprensione del codice**: Essendo conosciuti e avendo un nome inequivocabile, è possibile riconoscere i design pattern e creare un linguaggio comune quando si documenta e sviluppa il software.
+- **Soluzioni provate e testate**: Nascendo dall'esperienza congiunta di più sviluppatori, e avvalorati nel corso degli anni, i design pattern difficilmente nascondono delle sorprese, e la loro adozione può garantire una soluzione _battle-tested_ e dimostrata ad un problema specifico.
+
+Naturalmente, come ogni cosa nel mondo del software, i design pattern comportano degli svantaggi, tra cui:
+
+1. Possono generare parti di codice più complicate. Una porzione di codice intelligibile e senza logiche complesse potrebbe beneficiare dal non utilizzo dei design pattern, che ne complicherebbero la struttura. 
```

Comment:
```suggestion
1. Può portare ad aumentare la complessità del codice, per cui andrebbero utilizzati con parsimonia per non creare sovrastrutture non necessarie. 
```

Nota a margine: i design pattern non introducono complessità, se non applicati male o in casi d'uso limite. Non lo metterei tra gli svantaggi, in quanto molto frutto dell'esperienza professionale e personale, ma chiaramente è la mia opinione!

Reply:
Qui in generale è opinabile sia non metterlo che metterlo quindi ne possiamo parlare tranquillamente.
In generale ho letto che a volte si può overingegnerizzare utilizzandoli quando magari effettivamente non servono nel caso specifico , non so se questo può considerarsi un utilizzo sbagliato però ci si può ragionare

Reply:
Credo che il suggerimento di Serena sia coerente, credo sia il giusto compromesso tra "usarli non crea problemi" e "occhio a non abusarne"